### PR TITLE
Photon and ValueMap

### PR DESCRIPTION
**Pull request by nsmith-**
I took the seed cluster from the photon and passed it through the tool, and numbers came out!
I'll let my ntupler run for a bit to see how they look, but also here is a ValueMap implementation that could be used for MINIAOD when you integrate this into cmssw.  I let the electron implementation to you ;)

### CODE REVIEW

**Comment by artlbv** [comment]
Hi Nick, great that it works! And thanks for the value map — we’check that for the electrons as well.
However, I don’t think it’s a good idea to use the same ID helper for electrons and photons as now: the track variables are very different and mixing the accessors does not seem to be safe to me. 
What do you think?

**Comment by nsmith-** [comment]
Well yeah, storing an electron pointer and then not checking null is asking for trouble.
As I mentioned in a commit message, perhaps ElectronIDHelper could be called EgammaIDHelper or something and the accessors can have a `nullptr` guard around them in case it is not an electron in the current iteration.

**Comment by jkiesele** [comment]
I agree with Nick, calling it EgammaIDHelper makes more sense in this case.
But for now, I would just add the guards. I can do a refactoring PR after that if you want.

File: EgammaAnalysis/interface/LongDeps.h [hunk 1]
```diff
@@ -14,6 +14,7 @@ class LongDeps{
         inline float energyFH() const {return energyFH_;}
         inline float energyBH() const {return energyBH_;}
         inline const std::vector<float>& energyPerLayer() const {return energyPerLayer_;}
+        inline float energy(size_t layer) const {return (layer<layers_.size()) ? energyPerLayer_[layer] : 0.;}
```
Comment (artlbv) [inline EgammaAnalysis/interface/LongDeps.h]: Why this extra member? `energyPerLayer_` is initialized with 0s for all layers, and the layers.size check is incorrect: e.g. the cluster can contain layer 30 while spanning only 20 layers.

Comment (nsmith-) [reply EgammaAnalysis/interface/LongDeps.h]: The function signature in VARIABLES.md was something like this, so I assumed it was just missing. By all means please fix this as necessary, I was just guessing.

Comment (nsmith-) [reply EgammaAnalysis/interface/LongDeps.h]: I guess it should be `energyPerLayer_.size()`

File: EgammaAnalysis/plugins/HGCalPhotonIDValueMapProducer.cc [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+      // Fill here all the ValueMaps from their appropriate functions
+      maps_["sigmaUU"].push_back(eIDHelper_->sigmaUU());
+      maps_["sigmaVV"].push_back(eIDHelper_->sigmaVV());
+      maps_["sigmaEE"].push_back(eIDHelper_->sigmaEE());
+      maps_["sigmaPP"].push_back(eIDHelper_->sigmaPP());
+      maps_["nLayers"].push_back(ld.nLayers());
+      maps_["firstLayer"].push_back(ld.firstLayer());
+      maps_["lastLayer"].push_back(ld.lastLayer());
+      maps_["firstLayerEnergy"].push_back(ld.energy(ld.firstLayer()));
```
Comment (nsmith-) [inline EgammaAnalysis/plugins/HGCalPhotonIDValueMapProducer.cc]: So is this correct?

**Comment by jkiesele** [comment]
Hi all,
is there anything else open in this PR?

**Comment by jkiesele** [comment]
Dear all,
after some research, I unfortunately just discovered that the map as implemented here will not work as it is for the miniAOD production. But I found an example of how it should be implemented and it should not require too many changes:
https://insight.io/github.com/cms-sw/cmssw/blob/master/RecoEgamma/PhotonIdentification/plugins/PhotonIDValueMapProducer.cc

In particular this line will be helpful:
https://insight.io/github.com/cms-sw/cmssw/blob/master/RecoEgamma/PhotonIdentification/plugins/PhotonIDValueMapProducer.cc?line=443

File: EgammaAnalysis/plugins/HGCalPhotonIDValueMapProducer.cc [hunk 3]
```diff
@@ -67,6 +67,17 @@ HGCalPhotonIDValueMapProducer::HGCalPhotonIDValueMapProducer(const edm::Paramete
   maps_["sigmaVV"] = {};
   maps_["sigmaEE"] = {};
   maps_["sigmaPP"] = {};
+  maps_["nLayers"] = {};
+  maps_["firstLayer"] = {};
+  maps_["lastLayer"] = {};
+  maps_["firstLayerEnergy"] = {};
```
Comment (beaudett) [inline EgammaAnalysis/plugins/HGCalPhotonIDValueMapProducer.cc]: Hello, 
sorry for the late reaction. I am not sure that "firstLayerEnergy" is a very good variable for ID. Layer per layer energies can fluctuate (especially in the first layer for photons), and I suspect it is quite sensitive to PU. 
F.

File: EgammaAnalysis/src/PhotonHGCalIsoProducer.cc [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+        const size_t& ring=ringasso_.at(layer);
+
+        //do not consider hits associated to the photon cluster
+        if(std::find(photonhits.begin(),photonhits.end(),hit.first)==photonhits.end()){
+            isoringdeposits_.at(ring)+=hit.second->energy();
+        }
+    }
+
+    //does this make sense here or should it be done later?
```
Comment (nsmith-) [inline EgammaAnalysis/src/PhotonHGCalIsoProducer.cc]: @jkiesele I think leaving the raw energy might be safer than the ratio.  For example, the run 2 photon isolation cuts are only weakly pt-dependent.  Anyway the photon energy stored in that variable could be at some point a regression energy or some other thing that is not just sum(rechits).

Comment (jkiesele) [reply EgammaAnalysis/src/PhotonHGCalIsoProducer.cc]: Fully agree with you.

**Comment by artlbv** [comment]
Ok, looks good to me. Will try to add @beaudett's electron value map soon as well.

**Pull request merged**